pyodk/_endpoints/entities.py

```diff
@@ -0,0 +1,196 @@
+from pyodk._utils.session import Session
+from pyodk.errors import PyODKError
+
+log = logging.getLogger(__name__)
+
+
+class CurrentVersion(bases.Model):
+    label: str
+    current: bool
+    creatorId: int
+    userAgent: str
+    version: int
+    baseVersion: int | None = None
+    conflictingProperties: list[str] | None = None
+
+
+class Entity(bases.Model):
+    uuid: str
+    creatorId: int
+    createdAt: datetime
+    currentVersion: CurrentVersion
+    updatedAt: datetime | None = None
+    deletedAt: datetime | None = None
+
+
+class URLs(bases.Model):
+    class Config:
+        frozen = True
+
+    _entity_name: str = "projects/{project_id}/datasets/{el_name}"
+    list: str = f"{_entity_name}/entities"
+    post: str = f"{_entity_name}/entities"
+    get_table: str = f"{_entity_name}.svc/Entities"
+
+
+class EntityService(bases.Service):
+    """
+    Entity-related functionality is accessed through `client.entities`. For example:
+
+    ```python
+    from pyodk.client import Client
+
+    client = Client()
+    data = client.entities.list()
+    ```
+
+    Conceptually, an Entity's parent object is an EntityList. Each EntityList may
+    have multiple Entities. In Python parlance, EntityLists are like classes, while
+    Entities are like instances.
+    """
+
+    __slots__ = ("urls", "session", "default_project_id", "default_entity_list_name")
+
+    def __init__(
+        self,
+        session: Session,
+        default_project_id: int | None = None,
+        default_entity_list_name: str | None = None,
+        urls: URLs = None,
+    ):
+        self.urls: URLs = urls if urls is not None else URLs()
+        self.session: Session = session
+        self.default_project_id: int | None = default_project_id
+        self.default_entity_list_name: str | None = default_entity_list_name
+
+    def list(
+        self, entity_list_name: str | None = None, project_id: int | None = None
+    ) -> list[Entity]:
+        """
+        Read all Entity metadata.
+
+        :param entity_list_name: The name of the Entity List (Dataset) being referenced.
+        :param project_id: The id of the project the Entity belongs to.
+
+        :return: A list of the object representation of all Entity metadata.
+        """
+        try:
+            pid = pv.validate_project_id(project_id, self.default_project_id)
+            eln = pv.validate_entity_list_name(
+                entity_list_name, self.default_entity_list_name
+            )
+        except PyODKError as err:
+            log.error(err, exc_info=True)
+            raise
+
+        response = self.session.response_or_error(
+            method="GET",
+            url=self.session.urlformat(self.urls.list, project_id=pid, el_name=eln),
+            logger=log,
+        )
+        data = response.json()
+        return [Entity(**r) for r in data]
+
+    def create(
+        self,
+        label: str,
+        data: dict,
+        entity_list_name: str | None = None,
```

Comment:
This should also be required. There is no notion of an entity that lives outside of an entity list.

Reply:
Ah, right, it's because a default can be set. I seem to have trouble remembering that pattern!

Reply:
Yep!